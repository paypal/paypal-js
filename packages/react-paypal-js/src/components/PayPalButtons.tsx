import React, { useEffect, useRef, useState } from "react";

import { usePayPalScriptReducer } from "../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace, generateErrorMessage } from "../utils";
import { SDK_SETTINGS } from "../constants";

import type { FunctionComponent } from "react";
import type { PayPalButtonsComponent, OnInitActions } from "@paypal/paypal-js";
import type { PayPalButtonsComponentProps } from "../types";

/**
This `<PayPalButtons />` component supports rendering [buttons](https://developer.paypal.com/docs/business/javascript-sdk/javascript-sdk-reference/#buttons) for PayPal, Venmo, and alternative payment methods.
It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.
*/
export const PayPalButtons: FunctionComponent<PayPalButtonsComponentProps> = ({
    className = "",
    disabled = false,
    children,
    forceReRender = [],
    ...buttonProps
}: PayPalButtonsComponentProps) => {
    const isDisabledStyle = disabled ? { opacity: 0.38 } : {};
    const classNames = `${className} ${
        disabled ? "paypal-buttons-disabled" : ""
    }`.trim();
    const buttonsContainerRef = useRef<HTMLDivElement>(null);
    const buttons = useRef<PayPalButtonsComponent | null>(null);

    const [{ isResolved, options }] = usePayPalScriptReducer();
    const [initActions, setInitActions] = useState<OnInitActions | null>(null);
    const [isEligible, setIsEligible] = useState(true);
    const [, setErrorState] = useState(null);

    function closeButtonsComponent() {
        if (buttons.current !== null) {
            buttons.current.close().catch(() => {
                // ignore errors when closing the component
            });
        }
    }

    // useEffect hook for rendering the buttons
    useEffect(() => {
        // verify the sdk script has successfully loaded
        if (isResolved === false) {
            return closeButtonsComponent;
        }

        const paypalWindowNamespace = getPayPalWindowNamespace(
            options.dataNamespace
        );

        // verify dependency on window object
        if (
            paypalWindowNamespace === undefined ||
            paypalWindowNamespace.Buttons === undefined
        ) {
            setErrorState(() => {
                throw new Error(
                    generateErrorMessage({
                        reactComponentName: PayPalButtons.displayName as string,
                        sdkComponentKey: "buttons",
                        sdkRequestedComponents: options.components,
                        sdkDataNamespace: options[SDK_SETTINGS.DATA_NAMESPACE],
                    })
                );
            });
            return closeButtonsComponent;
        }

        const decoratedOnInit = (
            data: Record<string, unknown>,
            actions: OnInitActions
        ) => {
            setInitActions(actions);
            if (typeof buttonProps.onInit === "function") {
                buttonProps.onInit(data, actions);
            }
        };

        try {
            buttons.current = paypalWindowNamespace.Buttons({
                ...buttonProps,
                onInit: decoratedOnInit,
            });
        } catch (err) {
            return setErrorState(() => {
                throw new Error(
                    `Failed to render <PayPalButtons /> component. Failed to initialize:  ${err}`
                );
            });
        }

        // only render the button when eligible
        if (buttons.current.isEligible() === false) {
            setIsEligible(false);
            return closeButtonsComponent;
        }

        if (!buttonsContainerRef.current) {
            return closeButtonsComponent;
        }

        buttons.current.render(buttonsContainerRef.current).catch((err) => {
            // component failed to render, possibly because it was closed or destroyed.
            if (
                buttonsContainerRef.current === null ||
                buttonsContainerRef.current.children.length === 0
            ) {
                // paypal buttons container is no longer in the DOM, we can safely ignore the error
                return;
            }
            // paypal buttons container is still in the DOM
            setErrorState(() => {
                throw new Error(
                    `Failed to render <PayPalButtons /> component. ${err}`
                );
            });
        });

        return closeButtonsComponent;
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isResolved, ...forceReRender, buttonProps.fundingSource]);

    // useEffect hook for managing disabled state
    useEffect(() => {
        if (initActions === null) {
            return;
        }

        if (disabled === true) {
            initActions.disable().catch(() => {
                // ignore errors when disabling the component
            });
        } else {
            initActions.enable().catch(() => {
                // ignore errors when enabling the component
            });
        }
    }, [disabled, initActions]);

    return (
        <>
            {isEligible ? (
                <div
                    ref={buttonsContainerRef}
                    style={isDisabledStyle}
                    className={classNames}
                />
            ) : (
                children
            )}
        </>
    );
};

PayPalButtons.displayName = "PayPalButtons";
