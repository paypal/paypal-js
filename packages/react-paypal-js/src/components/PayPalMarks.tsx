import React, { useEffect, useRef, useState } from "react";

import { usePayPalScriptReducer } from "../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace, generateErrorMessage } from "../utils";
import { SDK_SETTINGS } from "../constants";

import type { FC, ReactNode } from "react";
import type {
    PayPalMarksComponentOptions,
    PayPalMarksComponent,
} from "@paypal/paypal-js";

export interface PayPalMarksComponentProps extends PayPalMarksComponentOptions {
    /**
     * Pass a css class to the div container.
     */
    className?: string;
    children?: ReactNode;
}

/**
The `<PayPalMarks />` component is used for conditionally rendering different payment options using radio buttons.
The [Display PayPal Buttons with other Payment Methods guide](https://developer.paypal.com/docs/business/checkout/add-capabilities/buyer-experience/#display-paypal-buttons-with-other-payment-methods) describes this style of integration in detail.
It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.

This component can also be configured to use a single funding source similar to the [standalone buttons](https://developer.paypal.com/docs/business/checkout/configure-payments/standalone-buttons/) approach.
A `FUNDING` object is exported by this library which has a key for every available funding source option.
*/
export const PayPalMarks: FC<PayPalMarksComponentProps> = ({
    className = "",
    children,
    ...markProps
}: PayPalMarksComponentProps) => {
    const [{ isResolved, options }] = usePayPalScriptReducer();
    const markContainerRef = useRef<HTMLDivElement>(null);
    const [isEligible, setIsEligible] = useState(true);
    const [, setErrorState] = useState(null);

    /**
     * Render PayPal Mark into the DOM
     */
    const renderPayPalMark = (mark: PayPalMarksComponent) => {
        const { current } = markContainerRef;

        // only render the mark when eligible
        if (!current || !mark.isEligible()) {
            return setIsEligible(false);
        }
        // Remove any children before render it again
        if (current.firstChild) {
            current.removeChild(current.firstChild);
        }

        mark.render(current).catch((err) => {
            // component failed to render, possibly because it was closed or destroyed.
            if (current === null || current.children.length === 0) {
                // paypal marks container is no longer in the DOM, we can safely ignore the error
                return;
            }
            // paypal marks container is still in the DOM
            setErrorState(() => {
                throw new Error(
                    `Failed to render <PayPalMarks /> component. ${err}`
                );
            });
        });
    };

    useEffect(() => {
        // verify the sdk script has successfully loaded
        if (isResolved === false) {
            return;
        }

        const paypalWindowNamespace = getPayPalWindowNamespace(
            options[SDK_SETTINGS.DATA_NAMESPACE]
        );

        // verify dependency on window object
        if (
            paypalWindowNamespace === undefined ||
            paypalWindowNamespace.Marks === undefined
        ) {
            return setErrorState(() => {
                throw new Error(
                    generateErrorMessage({
                        reactComponentName: PayPalMarks.displayName as string,
                        sdkComponentKey: "marks",
                        sdkRequestedComponents: options.components,
                        sdkDataNamespace: options[SDK_SETTINGS.DATA_NAMESPACE],
                    })
                );
            });
        }

        renderPayPalMark(paypalWindowNamespace.Marks({ ...markProps }));
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isResolved, markProps.fundingSource]);

    return (
        <>
            {isEligible ? (
                <div ref={markContainerRef} className={className} />
            ) : (
                children
            )}
        </>
    );
};

PayPalMarks.displayName = "PayPalMarks";
