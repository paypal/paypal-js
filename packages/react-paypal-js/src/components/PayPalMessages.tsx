import React, { useEffect, useRef, useState } from "react";

import { usePayPalScriptReducer } from "../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace, generateErrorMessage } from "../utils";
import { SDK_SETTINGS } from "../constants";

import type { FC } from "react";
import type {
    PayPalMessagesComponentOptions,
    PayPalMessagesComponent,
} from "@paypal/paypal-js";

export interface PayPalMessagesComponentProps
    extends PayPalMessagesComponentOptions {
    forceReRender?: unknown[];
    className?: string;
}

/**
This `<PayPalMessages />` messages component renders a credit messaging on upstream merchant sites.
It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.
*/
export const PayPalMessages: FC<PayPalMessagesComponentProps> = ({
    className = "",
    forceReRender = [],
    ...messageProps
}: PayPalMessagesComponentProps) => {
    const [{ isResolved, options }] = usePayPalScriptReducer();
    const messagesContainerRef = useRef<HTMLDivElement>(null);
    const messages = useRef<PayPalMessagesComponent | null>(null);
    const [, setErrorState] = useState(null);

    useEffect(() => {
        // verify the sdk script has successfully loaded
        if (isResolved === false) {
            return;
        }

        const paypalWindowNamespace = getPayPalWindowNamespace(
            options[SDK_SETTINGS.DATA_NAMESPACE]
        );

        // verify dependency on window object
        if (
            paypalWindowNamespace === undefined ||
            paypalWindowNamespace.Messages === undefined
        ) {
            return setErrorState(() => {
                throw new Error(
                    generateErrorMessage({
                        reactComponentName:
                            PayPalMessages.displayName as string,
                        sdkComponentKey: "messages",
                        sdkRequestedComponents: options.components,
                        sdkDataNamespace: options[SDK_SETTINGS.DATA_NAMESPACE],
                    })
                );
            });
        }

        messages.current = paypalWindowNamespace.Messages({
            ...messageProps,
        });

        messages.current
            .render(messagesContainerRef.current as HTMLDivElement)
            .catch((err) => {
                // component failed to render, possibly because it was closed or destroyed.
                if (
                    messagesContainerRef.current === null ||
                    messagesContainerRef.current.children.length === 0
                ) {
                    // paypal messages container is no longer in the DOM, we can safely ignore the error
                    return;
                }
                // paypal messages container is still in the DOM
                setErrorState(() => {
                    throw new Error(
                        `Failed to render <PayPalMessages /> component. ${err}`
                    );
                });
            });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isResolved, ...forceReRender]);

    return <div ref={messagesContainerRef} className={className} />;
};

PayPalMessages.displayName = "PayPalMessages";
