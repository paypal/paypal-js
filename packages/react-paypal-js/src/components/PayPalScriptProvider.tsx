import React, { useEffect, useReducer } from "react";
import { loadScript } from "@paypal/paypal-js";

import {
    getScriptID,
    ScriptContext,
    scriptReducer,
} from "../context/scriptProviderContext";
import { SCRIPT_ID, SDK_SETTINGS, LOAD_SCRIPT_ERROR } from "../constants";
import { SCRIPT_LOADING_STATE, DISPATCH_ACTION } from "../types";

import type { FC } from "react";
import type { ScriptProviderProps } from "../types";

/**
This `<PayPalScriptProvider />` component takes care of loading the JS SDK `<script>`.
It manages state for script loading so children components like `<PayPalButtons />` know when it's safe to use the `window.paypal` global namespace.

Note: You always should use this component as a wrapper for  `PayPalButtons`, `PayPalMarks`, `PayPalMessages` and `BraintreePayPalButtons` components.
 */
export const PayPalScriptProvider: FC<ScriptProviderProps> = ({
    options = { clientId: "test" },
    children,
    deferLoading = false,
}: ScriptProviderProps) => {
    const [state, dispatch] = useReducer(scriptReducer, {
        options: {
            ...options,
            [SDK_SETTINGS.DATA_JS_SDK_LIBRARY]: SDK_SETTINGS.DATA_LIBRARY_VALUE,
            [SDK_SETTINGS.DATA_SDK_INTEGRATION_SOURCE]:
                SDK_SETTINGS.DATA_LIBRARY_VALUE,
            [SCRIPT_ID]: `${getScriptID(options)}`,
        },
        loadingStatus: deferLoading
            ? SCRIPT_LOADING_STATE.INITIAL
            : SCRIPT_LOADING_STATE.PENDING,
    });

    useEffect(() => {
        if (
            deferLoading === false &&
            state.loadingStatus === SCRIPT_LOADING_STATE.INITIAL
        ) {
            return dispatch({
                type: DISPATCH_ACTION.LOADING_STATUS,
                value: SCRIPT_LOADING_STATE.PENDING,
            });
        }

        if (state.loadingStatus !== SCRIPT_LOADING_STATE.PENDING) {
            return;
        }

        let isSubscribed = true;

        loadScript(state.options)
            .then(() => {
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: SCRIPT_LOADING_STATE.RESOLVED,
                    });
                }
            })
            .catch((err) => {
                console.error(`${LOAD_SCRIPT_ERROR} ${err}`);
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: {
                            state: SCRIPT_LOADING_STATE.REJECTED,
                            message: String(err),
                        },
                    });
                }
            });
        return () => {
            isSubscribed = false;
        };
    }, [state.options, deferLoading, state.loadingStatus]);

    return (
        <ScriptContext.Provider value={{ ...state, dispatch }}>
            {children}
        </ScriptContext.Provider>
    );
};
