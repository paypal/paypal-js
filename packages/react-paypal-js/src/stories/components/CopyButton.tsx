import React, { useState } from "react";
import { useSandpack } from "@codesandbox/sandpack-react";

import type { ReactElement } from "react";
import type {
    FontWeightProperty,
    FloatProperty,
    PositionProperty,
} from "csstype/index";

const COPY_BUTTON = {
    padding: "4px 10px",
    cursor: "pointer",
    display: "flex",
    alignItems: "center",
    color: "#333333",
    background: "#FFFFFF",
    fontSize: "12px",
    lineHeight: "16px",
    fontWeight: "700" as FontWeightProperty,
    border: "1px solid rgba(0,0,0,.1)",
    borderRadius: "4px 0 0 0",
    float: "right" as FloatProperty,
    position: "relative" as PositionProperty,
    top: "-26px",
};

/**
 * This component is use to copy the behavior from the Description component
 * in the storybook Doc pages. It use the state of the Sandpack library
 * to get the code inside the App.js file and write it into the clipboard
 *
 * @returns a ReactElement with the custom copy button
 */
const CopyButton = (): ReactElement => {
    const { sandpack } = useSandpack();
    const [isCopy, setIsCopy] = useState<boolean>(false);

    const getCopyBorder = () => (isCopy ? "2px solid #1DA7FD" : "none");

    const onClickHandler = () => {
        navigator.clipboard.writeText(sandpack.files[sandpack.activePath].code);
        setIsCopy(true);
        // Same behavior from default Description storybook component
        setTimeout(() => {
            setIsCopy(false);
        }, 1000);
    };

    return (
        <button
            onClick={onClickHandler}
            style={{ ...COPY_BUTTON, borderBottom: getCopyBorder() }}
        >
            {isCopy ? "Copied" : "Copy"}
        </button>
    );
};

export default CopyButton;
