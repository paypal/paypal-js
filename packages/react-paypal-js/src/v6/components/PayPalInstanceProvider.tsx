import React, { useEffect, useState } from "react";
import { loadCoreSdkScript } from "@paypal/paypal-js/sdk-v6";

import { InstanceContext } from "../context/InstanceProviderContext";
import { isServer } from "../../utils";

import type {
    CreateInstanceOptions,
    Components,
    SdkInstance,
    EligiblePaymentMethodsOutput,
    LoadCoreSdkScriptOptions,
} from "../types";

interface PayPalInstanceProviderProps {
    options: CreateInstanceOptions<readonly [Components, ...Components[]]>;
    children: React.ReactNode;
    scriptOptions: LoadCoreSdkScriptOptions;
}

export const PayPalInstanceProvider: React.FC<PayPalInstanceProviderProps> = ({
    options,
    children,
    scriptOptions,
}: {
    options: CreateInstanceOptions<readonly [Components, ...Components[]]>;
    children: React.ReactNode;
    scriptOptions: LoadCoreSdkScriptOptions;
}) => {
    const [sdkInstance, setSdkInstance] = useState<SdkInstance<
        readonly [Components, ...Components[]]
    > | null>(null);
    const [eligiblePaymentMethods, setEligiblePaymentMethods] =
        useState<EligiblePaymentMethodsOutput | null>(null);
    const [isLoading, setIsLoading] = useState(true);
    const [error, setError] = useState<Error | null>(null);

    useEffect(() => {
        // Skip if SSR
        if (isServer) {
            setIsLoading(false);
            return;
        }

        const controller = new AbortController();

        const initializeSdk = async () => {
            try {
                setIsLoading(true);
                setError(null);

                const paypalNameSapce = await loadCoreSdkScript(scriptOptions);

                if (controller.signal.aborted) {
                    return;
                }

                if (!paypalNameSapce) {
                    throw new Error("PayPal SDK failed to load");
                }

                const instance = await paypalNameSapce.createInstance(options);

                if (controller.signal.aborted) {
                    return;
                }

                setSdkInstance(instance);

                try {
                    const eligiblePaymentMethods =
                        await instance.findEligibleMethods({});

                    if (controller.signal.aborted) {
                        return;
                    }

                    setEligiblePaymentMethods(eligiblePaymentMethods);
                } catch (eligibilityErr) {
                    if (!controller.signal.aborted) {
                        console.warn(
                            "Failed to get eligible payment methods:",
                            eligibilityErr,
                        );
                    }
                }

                if (!controller.signal.aborted) {
                    setIsLoading(false);
                }
            } catch (err) {
                if (!controller.signal.aborted) {
                    const error =
                        err instanceof Error ? err : new Error(String(err));
                    setError(error);
                    setIsLoading(false);
                }
            }
        };

        initializeSdk();

        return () => {
            controller.abort();
        };
    }, [options, scriptOptions]);

    return (
        <InstanceContext.Provider
            value={{ sdkInstance, eligiblePaymentMethods, isLoading, error }}
        >
            {children}
        </InstanceContext.Provider>
    );
};
