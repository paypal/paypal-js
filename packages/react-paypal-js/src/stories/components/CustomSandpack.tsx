import React from "react";
import {
    SandpackProvider,
    SandpackThemeProvider,
    SandpackPreview,
    SandpackCodeEditor,
} from "@codesandbox/sandpack-react";

import CopyButton from "./CopyButton";

import type { ReactElement } from "react";
import type {
    SandpackPredefinedTemplate,
    SandpackFile,
} from "@codesandbox/sandpack-react/dist/types/types";
import "@codesandbox/sandpack-react/dist/index.css";

// Default styles
const CUSTOM_STYLE = {
    width: "100%",
    height: "450px",
    border: "1px solid rgba(0,0,0,.1)",
    boxShadow: "rgb(0 0 0 / 10%) 0 1px 3px 0",
};

/**
 * Custom component to adjust the Sandpack react library UI to the
 * storybook requirements for a better user experience
 *
 * @param props to modify the UI behavior and source files
 * @returns a ReactElement with the customized Sandpack component
 */
const CustomSandpack = (props: {
    template?: SandpackPredefinedTemplate;
    files?: Record<string, SandpackFile>;
    customSetup?: Record<string, unknown>;
    showLineNumbers?: boolean;
    previewHeight?: string;
    codeHeight?: string;
}): ReactElement => (
    <SandpackProvider template={props.template} customSetup={props.customSetup}>
        <SandpackThemeProvider>
            <SandpackPreview
                customStyle={{
                    ...CUSTOM_STYLE,
                    height: props.previewHeight,
                    padding: "1em",
                    borderRadius: "4px 4px 0 0",
                }}
            />
            <SandpackCodeEditor
                customStyle={{
                    ...CUSTOM_STYLE,
                    height: props.codeHeight,
                    borderRadius: "0 0 4px 4px",
                }}
                showLineNumbers={props.showLineNumbers || false}
            />
            <CopyButton />
        </SandpackThemeProvider>
    </SandpackProvider>
);

export default CustomSandpack;
