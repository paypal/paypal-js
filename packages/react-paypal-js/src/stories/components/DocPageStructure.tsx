import React from "react";
import {
    DocsContainer,
    Title,
    Subtitle,
    Description,
    ArgsTable,
    CURRENT_SELECTION,
} from "@storybook/addon-docs";
import dedent from "ts-dedent";

import CustomSandpack from "./CustomSandpack";
import { SANDPACK_STYLES } from "../constants";

import type { DocsContextProps } from "@storybook/addon-docs/dist/ts3.9/blocks";

/**
 * Component to override the default DocPage structure from the storybook
 *
 * @param param custom props and context
 * @returns an Element with the DocPage new structure
 */
const DocPageStructure = ({
    context,
    code = "",
    options = { previewHeight: "450px", codeHeight: "600px" },
}: {
    context: DocsContextProps;
    code: string;
    options?: { previewHeight: string; codeHeight: string };
}): JSX.Element => (
    <DocsContainer context={context}>
        <Title />
        <Subtitle />
        <Description />
        <CustomSandpack
            template="react"
            customSetup={{
                dependencies: {
                    "@paypal/react-paypal-js": "latest",
                },
                entry: "/index.js",
                files: {
                    "/App.js": dedent`${code}`,
                    "/styles.css": {
                        code: dedent`${SANDPACK_STYLES}`,
                        hidden: true,
                    },
                },
            }}
            showLineNumbers={true}
            previewHeight={options?.previewHeight || "450px"}
            codeHeight={options?.codeHeight || "600px"}
        />
        <ArgsTable story={CURRENT_SELECTION} />
    </DocsContainer>
);

export default DocPageStructure;
