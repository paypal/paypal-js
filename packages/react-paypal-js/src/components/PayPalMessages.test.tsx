import React from "react";
import { render, waitFor } from "@testing-library/react";
import { mock } from "jest-mock-extended";
import { loadScript } from "@paypal/paypal-js";
import { ErrorBoundary } from "react-error-boundary";

import { PayPalMessages } from "..";
import { PayPalScriptProvider } from "../components/PayPalScriptProvider";

import type { ReactNode } from "react";
import type { PayPalNamespace } from "@paypal/paypal-js";

jest.mock("@paypal/paypal-js", () => ({
    loadScript: jest.fn(),
}));
const mockPayPalNamespace = mock<PayPalNamespace>();
const onError = jest.fn();
const wrapper = ({ children }: { children: ReactNode }) => (
    <ErrorBoundary fallback={<div>Error</div>} onError={onError}>
        {children}
    </ErrorBoundary>
);

describe("<PayPalMessages />", () => {
    beforeEach(() => {
        window.paypal = mockPayPalNamespace;
        (loadScript as jest.Mock).mockResolvedValue(mockPayPalNamespace);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("should render PayPalMessages with props", async () => {
        const style = {
            color: "blue",
        };
        window.paypal = {
            Messages: jest.fn(() => ({
                render: jest.fn().mockResolvedValue({}),
            })),
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{ clientId: "test", components: "messages" }}
            >
                <PayPalMessages style={style} className="customClass" />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(window.paypal?.Messages).toHaveBeenCalledWith({ style })
        );
    });

    test("should use className prop and add to div container", async () => {
        window.paypal = {
            Messages: jest.fn(() => ({
                render: jest.fn().mockResolvedValue({}),
            })),
            version: "",
        };

        const { container } = render(
            <PayPalScriptProvider
                options={{ clientId: "test", components: "messages" }}
            >
                <PayPalMessages className="custom-class-name" />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(
                container.querySelector("div.custom-class-name")
            ).not.toBeNull()
        );
    });

    test("should throw an error when namespace is undefined", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal = undefined;

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalMessages />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });

    test("should throw an error when the 'messages' function is not available in the namespace", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal = {
            Messages: undefined,
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test",
                    components: "messages",
                }}
            >
                <PayPalMessages />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });

    test("should throw an error when the render process fails and ignore error if the container is no longer in the DOM", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        const mockRender = jest
            .fn()
            .mockRejectedValue(new Error("Fail to render"));
        window.paypal = {
            Messages() {
                return { render: mockRender };
            },
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test",
                    components: "messages",
                }}
            >
                <PayPalMessages />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(mockRender).toBeCalled());
        spyConsoleError.mockRestore();
    });

    test("should throw an error when the render process fails and the container is in the DOM", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        const mockRender = jest.fn((element) => {
            // Add a children component to avoid ignoring error
            const markElement = document.createElement("div");

            markElement.setAttribute("id", "children-div");
            element.append(markElement);
            return Promise.reject();
        });
        window.paypal = {
            Messages() {
                return { render: mockRender };
            },
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test",
                    components: "messages",
                }}
            >
                <PayPalMessages />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });
});
