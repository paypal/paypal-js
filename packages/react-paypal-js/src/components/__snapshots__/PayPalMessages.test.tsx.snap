// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PayPalMessages /> should throw an error when namespace is undefined 1`] = `
"Unable to render <PayPalMessages /> because window.paypal.Messages is undefined.
To fix the issue, add 'messages' to the list of components passed to the parent PayPalScriptProvider:
\`<PayPalScriptProvider options={{ components: 'messages'}}>\`."
`;

exports[`<PayPalMessages /> should throw an error when the 'messages' function is not available in the namespace 1`] = `"Unable to render <PayPalMessages /> because window.paypal.Messages is undefined."`;

exports[`<PayPalMessages /> should throw an error when the render process fails and the container is in the DOM 1`] = `"Failed to render <PayPalMessages /> component. undefined"`;
